### PR TITLE
Allow editing actual posts and pages

### PR DESCRIPTION
**Pull request by nylen**
This PR builds the initial flow for editing actual posts and pages in the Gutenberg editor.  Until the editor is further along, the link to Gutenberg is an additional link on the posts/pages screens rather than a replacement for the default editor:

Much like the built-in "Quick Edit" link, this link is hidden if JavaScript is disabled.

The PR works as follows:

- Restructure the editor initialization code to accept a post object in the format returned by the WP REST API, with the expectation that we will use this API to build out the editor.
- If a `?post_id=X` value is set, perform an internal (server-side) WP REST API request to load the post with id `X`.  If that succeeds, pass this data to the editor; otherwise use the current "dummy" post content.
- Add links to the Edit screens for all post types to edit posts/pages in the Gutenberg editor, using the new `post_id` parameter.

This works well in my testing; however, in order to see post content in the editor, you will currently need to make sure that your post content has some block comment delimiters.  For example:

```html
<!-- wp:core/text -->
Welcome to WordPress. This is your first post. Edit or delete it, then start writing!
<!-- /wp:core/text -->
```

Result:

2017-03-30t07 50 19-0700

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -47,6 +47,56 @@ function gutenberg_register_scripts() {
...
+			) ),
+			__( 'Gutenberg', 'gutenberg' )
+		);
+		// Insert the Gutenberg action immediately after the Edit action.
+		$edit_offset = array_search( 'edit', array_keys( $actions ), true );
+		$actions = array_merge(
+			array_slice( $actions, 0, $edit_offset + 1 ),
+			array( 'gutenberg hide-if-no-js' => $gutenberg_action ),
+			array_slice( $actions, $edit_offset + 1 )
+		);
```
Comment (nylen) [inline index.php]: This is a bit annoying, but `$actions` is an associative array and we need to insert a key into the middle of it.  Technique from http://stackoverflow.com/a/9847709/106302.

File: index.php [hunk 2]
```diff
@@ -56,16 +106,76 @@ function gutenberg_register_scripts() {
...
+	 */
+
+	// The editor code itself
+	wp_enqueue_script(
+		'wp-editor',
+		plugins_url( 'editor/build/index.js', __FILE__ ),
+		array( 'wp-blocks', 'wp-element', 'gutenberg-content' ),
+		false, // $ver
+		true   // $in_footer
+	);
```
Comment (nylen) [inline index.php]: Here and elsewhere, I've restructured the code in this function to avoid long lines, which are really detrimental to code readability.  I'd recommend that we do this throughout the project, and add rule(s) to our linting setup(s) accordingly.

File: index.php [hunk 3]
```diff
@@ -56,16 +106,76 @@ function gutenberg_register_scripts() {
...
  */
 function gutenberg_scripts_and_styles( $hook ) {
-	if ( 'toplevel_page_gutenberg' === $hook ) {
-		// Scripts
-		wp_register_script( 'gutenberg-content', plugins_url( 'post-content.js', __FILE__ ) );
-		wp_enqueue_script( 'wp-editor', plugins_url( 'editor/build/index.js', __FILE__ ), array( 'wp-blocks', 'wp-element', 'gutenberg-content' ), false, true );
-		wp_add_inline_script( 'wp-editor', 'wp.editor.createEditorInstance( \'editor\', { content: window.content } );' );
-
-		// Styles
-		wp_enqueue_style( 'wp-editor-font', 'https://fonts.googleapis.com/css?family=Noto+Serif:400,400i,700,700i', false );
```
Comment (nylen) [inline index.php]: The third `false` parameter here was incorrect.  This is the [`$deps` parameter](https://developer.wordpress.org/reference/functions/wp_enqueue_style/), which is [set to `array()` if it's not an array](https://github.com/WordPress/wordpress-develop/blob/4.7.3/src/wp-includes/class-wp-dependency.php#L83).

File: index.php [hunk 4]
```diff
@@ -56,16 +106,76 @@ function gutenberg_register_scripts() {
...
+			$post_to_edit = $response->get_data();
+		}
+	}
+
+	// Initialize the post data...
+	if ( $post_to_edit ) {
+		// ...with a real post
+		wp_add_inline_script(
+			'wp-editor',
+			'wp.editor.post = ' . json_encode( $post_to_edit )
```
Comment (aduth) [inline index.php]: The current approach with `wp.editor` assumes that multiple instances of an editor could be constructed, and the idea of a single `wp.editor.post` seems at odds with this. Obvious alternatives to me are:

- Stop supporting multiple instances of editor and refactor functions accordingly to operate off `wp.editor.post` directly
- Assign this as a separate global variable like `_wpGutenbergPost`

File: index.php [hunk 5]
```diff
@@ -56,16 +106,76 @@ function gutenberg_register_scripts() {
...
+	// Load an actual post if an ID is specified
+	$post_to_edit = null;
+	if ( isset( $_GET['post_id'] ) && (int) $_GET['post_id'] > 0 ) {
+		$request = new WP_REST_Request(
+			'GET',
+			sprintf( '/wp/v2/posts/%d', (int) $_GET['post_id'] )
+		);
+		$request->set_param( 'context', 'edit' );
+		$response = rest_do_request( $request );
+		if ( 200 === $response->get_status() ) {
```
Comment (aduth) [inline index.php]: Noting that we'll want to better handle the `else` condition here at some point (not found, not allowed).

**Review by aduth — APPROVED** [review]
Couple minor notes, but LGTM 👍

File: index.php (hunk 5 above)
Comment (nylen) [inline index.php]: I figured we weren't ready to do much about this yet.  Anything else you'd like to see in this PR?

File: index.php (hunk 4 above)
Comment (nylen) [inline index.php]: Why might we want to construct multiple instances of the editor?

Do we want to do that refactoring to a single editor instance now?  I'm guessing probably not, constructing an editor without depending on global state seems like a more desirable code structure regardless of how we end up using it.

If we do use a global variable we can use `wp_localize_script` which seems slightly more _preferred_ for injecting script data, but doesn't support object paths with `.` in them.

I don't have a strong preference here, but if we're going to change something on this PR, here's what I would do:

- Rename the current post to `_wpGutenbergPost` as you suggested
- Also rename the "default" test post content from `window.content` to something like `_wpGutenbergTestPost` (or, be slightly more clever and set this to `_wpGutenbergPost` if no post is loaded, simplifying the code path for initializing the editor).

Comment (aduth) [inline index.php]: >Why might we want to construct multiple instances of the editor?

One original consideration was that it simplifies swapping the existing implementation of [`wp_editor`](https://developer.wordpress.org/reference/functions/wp_editor/).

The approach you suggest seems reasonable to me.

File: index.php (hunk 5 above)
Comment (joyously) [inline index.php]: Load a placeholder editor-style.css.

Comment (nylen) [inline index.php]: @joyously I think this should be done in a separate PR.

**Comment by nylen** [comment]
Simplified the post initialization logic in @d032dc0.  In a future PR, I expect that error handling for invalid posts will make use of error responses returned from the REST API, and maybe use a custom error message if e.g. no post is specified.

I'm satisfied with this PR now.  Feel free to merge or leave further feedback which I will address after the weekend.

**Review by aduth — APPROVED** [review]
Looks good 👍

File: index.php (hunk 5 above)
Comment (nylen) [inline index.php]: @joyously if this is something you're interested in seeing sooner rather than later, can you please create an issue explaining a bit about what `editor-style.css` is used for and why we need to support it?
